src/config/loading/secret.rs

```diff
@@ -0,0 +1,341 @@
+use std::{collections::HashMap, io::Read};
+
+use futures::executor;
+use indexmap::IndexMap;
+use once_cell::sync::Lazy;
+use regex::{Captures, Regex};
+use serde::{Deserialize, Serialize};
+use tokio::{io::AsyncWriteExt, process::Command, time::Duration};
+use toml::value::Table;
+use typetag::serde;
+
+use super::{loader, prepare_input};
+use crate::config::{
+    loading::{deserialize_table, ComponentHint, Process},
+    ComponentKey,
+};
+use crate::signal;
+
+static COLLECTOR: Lazy<Regex> =
+    Lazy::new(|| Regex::new(r"SECRET\[([[:word:]]+)\.([[:word:].]+)\]").unwrap());
+
+#[typetag::serde(tag = "type")]
+pub trait SecretBackend: core::fmt::Debug + Send + Sync + dyn_clone::DynClone {
+    fn retrieve(
+        &mut self,
+        secret_keys: Vec<String>,
+        signal_rx: &mut signal::SignalRx,
+    ) -> crate::Result<HashMap<String, String>>;
+}
+
+#[derive(Deserialize, Serialize, Debug, Default)]
+pub struct SecretBackendLoader {
+    backends: IndexMap<ComponentKey, Box<dyn SecretBackend>>,
+    pub(crate) secret_keys: HashMap<String, Vec<String>>,
+}
+
+impl SecretBackendLoader {
+    pub(crate) fn new() -> Self {
+        Self {
+            backends: IndexMap::new(),
+            secret_keys: HashMap::new(),
+        }
+    }
+
+    pub(crate) fn retrieve(
+        &mut self,
+        signal_rx: &mut signal::SignalRx,
+    ) -> Result<HashMap<String, String>, String> {
+        let secrets = self.secret_keys.iter().flat_map(|(backend_name, keys)| {
+            match self.backends.get_mut(&ComponentKey::from(backend_name.clone())) {
+                None => {
+                    vec![Err(format!("Backend \"{}\" is required for secret retrieval but was not found in config.", backend_name))]
+                },
+                Some(backend) => {
+                    debug!(message = "Retrieving secret from a backend.", backend = ?backend_name);
+                    match backend.retrieve(keys.to_vec(), signal_rx) {
+                        Err(e) => {
+                            vec![Err(format!("Error while retrieving secret from backend \"{}\": {}.", backend_name, e))]
```

Comment:
I don't see where the errors are actually collected here to have all errors bubbled up, can you point out where that's happening - I guess the `vec!` on ln52 as well..

Reply:
cc/ @prognant

Reply:
Ah, thanks!